packages/hydrogen/src/logger.ts

```diff
@@ -0,0 +1,103 @@
+import {yellow, red} from 'kolorist';
+
+type LoggerMethod = (
+  context: {request: Request},
+  ...args: Array<any>
+) => void | Promise<any>;
+
+interface LoggerOptions {
+  trace: LoggerMethod;
+  debug: LoggerMethod;
+  warn: LoggerMethod;
+  error: LoggerMethod;
+  fatal: LoggerMethod;
+}
+
+export class Logger {
+  #waitUntil: (p: Promise<unknown>) => void;
+  #request: Request;
+
+  constructor(
+    waitUntil: (p: Promise<unknown>) => void,
+    request: Request,
+    options: Partial<LoggerOptions> = {},
+  ) {
+    for (const [key, value] of Object.entries(options)) {
+      this.defaultLogger[key as keyof LoggerOptions] = value;
+    }
+    this.#waitUntil = waitUntil;
+    this.#request = request;
+  }
+
+  trace(...args: Array<unknown>) {
+    this.#doLog('trace', ...args);
+  }
+
+  debug(...args: Array<unknown>) {
+    this.#doLog('debug', ...args);
+  }
+
+  warn(...args: Array<unknown>) {
+    this.#doLog('warn', ...args);
+  }
+
+  error(...args: Array<unknown>) {
+    this.#doLog('error', ...args);
+  }
+
+  fatal(...args: Array<unknown>) {
+    this.#doLog('fatal', ...args);
+  }
+
+  #doLog(method: keyof LoggerOptions, ...args: Array<any>) {
+    try {
+      const maybePromise = this.defaultLogger[method](
+        {request: this.#request},
+        ...args,
+      );
+      if (maybePromise instanceof Promise) {
+        this.#waitUntil(
+          maybePromise.catch((e) => {
+            const message = e instanceof Error ? e.stack : e;
+            console.error(
+              `Promise error from the custom logging implementation for logger.${method} failed:\n${message}`,
+            );
+          }),
+        );
+      }
+    } catch (e) {
+      const message = e instanceof Error ? e.stack : e;
```

Comment:
the `message` is from `e.stack`? or should it be from `e.message`?

Reply:
Yeah, the stack contains the message.